nifi-nar-bundles/nifi-graph-bundle/nifi-other-graph-services/src/main/java/org/apache/nifi/graph/GremlinBytecodeClientService.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.nifi.annotation.lifecycle.OnDisabled;
+import org.apache.nifi.annotation.lifecycle.OnEnabled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.expression.ExpressionLanguageScope;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.reporting.InitializationException;
+import org.apache.tinkerpop.gremlin.driver.Cluster;
+import org.apache.tinkerpop.gremlin.driver.remote.DriverRemoteConnection;
+import org.apache.tinkerpop.gremlin.process.traversal.AnonymousTraversalSource;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+
+import javax.script.Bindings;
+import javax.script.Compilable;
+import javax.script.CompiledScript;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class GremlinBytecodeClientService extends AbstractTinkerpopClientService implements GraphClientService {
+
+    private static final List<PropertyDescriptor> NEW_DESCRIPTORS;
+
+    public static final PropertyDescriptor ADDITIONAL_JARS = new PropertyDescriptor.Builder()
+            .name("gremlin-bytecode-additional-jars")
+            .displayName("Additional Jars")
+            .description("Additional jars needed for the traversal to work go here.")
+            .dynamicallyModifiesClasspath(true)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .build();
+
+    static {
+        List<PropertyDescriptor> _temp = new ArrayList<>();
+        _temp.addAll(DESCRIPTORS);
+        _temp.add(ADDITIONAL_JARS);
+        NEW_DESCRIPTORS = Collections.unmodifiableList(_temp);
+    }
+
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return NEW_DESCRIPTORS;
+    }
+
+    private ScriptEngineManager MANAGER = new ScriptEngineManager();
+    private ScriptEngine engine;
+    private Map<String, CompiledScript> compiledCode;
+    private Cluster cluster;
+
+    /**
+     * @param context
+     *            the configuration context
+     * @throws InitializationException
+     *             if unable to create a database connection
+     */
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) throws InitializationException {
+        cluster = buildCluster(context);
+
+        compiledCode = new ConcurrentHashMap<>();
+        engine = MANAGER.getEngineByName("groovy");
+    }
+
+    @OnDisabled
+    public void shutdown() {
+        try {
+            compiledCode = null;
+            engine = null;
+            cluster.close();
+        } catch (Exception e) {
+            throw new ProcessException(e);
+        }
+    }
+
```

Comment:
Would there ever be usage where s would be the same often?  If so, maybe caching the compiled code and executing it multiple times would be a good thing.  Maybe there could be an option to "cache byte code" or something?

If you don't expect that case, then never mind :)

Reply:
Yes, there would. Our graph APIs are built with the idea that you will parameterize your scripts because that's a best practice with either Cypher or Gremlin. This Gremlin process is a little different, but not so much that it would make any sense to skip parameterization.